dashboard/modules/reporter/reporter_agent.py

```diff
@@ -394,30 +397,66 @@ def _get_cpu_percent(in_k8s: bool):
     @staticmethod
     def _get_gpu_usage():
         global enable_gpu_usage_check
-        if gpustat is None or not enable_gpu_usage_check:
+        if not enable_gpu_usage_check:
             return []
         gpu_utilizations = []
-        gpus = []
         try:
-            gpus = gpustat.new_query().gpus
+            pynvml.nvmlInit()
+            num_gpus = pynvml.nvmlDeviceGetCount()
+            for i in range(num_gpus):
+                gpu_handle = pynvml.nvmlDeviceGetHandleByIndex(i)
+                utilization = None
+                try:
+                    utilization_info = pynvml.nvmlDeviceGetUtilizationRates(gpu_handle)
+                    utilization = int(utilization_info.gpu)
+                except pynvml.NVMLError as e:
+                    logger.debug(f"pynvml failed to retrieve GPU utilization: {e}")
+
+                # processes pids
+                processes_pids = None
+                try:
+                    nv_comp_processes = pynvml.nvmlDeviceGetComputeRunningProcesses(
+                        gpu_handle
+                    )
+                    nv_graphics_processes = (
+                        pynvml.nvmlDeviceGetGraphicsRunningProcesses(gpu_handle)
+                    )
+                    processes_pids = [
+                        ProcessGPUInfo(
+                            pid=nv_process.pid,
+                            gpu_memory_usage=nv_process.usedGpuMemory // MB
+                            if nv_process.usedGpuMemory
+                            else 0,
+                        )
+                        for nv_process in (nv_comp_processes + nv_graphics_processes)
+                    ]
+                except pynvml.NVMLError as e:
+                    logger.debug(f"pynvml failed to retrieve GPU processes: {e}")
+
+                info = GpuUtilizationInfo(
+                    index=i,
+                    name=pynvml.nvmlDeviceGetName(gpu_handle),
+                    uuid=pynvml.nvmlDeviceGetUUID(gpu_handle),
+                    utilization_gpu=utilization,
+                    memory_used=int(pynvml.nvmlDeviceGetMemoryInfo(gpu_handle).used)
+                    // MB,
+                    memory_total=int(pynvml.nvmlDeviceGetMemoryInfo(gpu_handle).total)
+                    // MB,
+                    processes_pids=processes_pids,
+                )
+                gpu_utilizations.append(info)
+            pynvml.nvmlShutdown()
         except Exception as e:
```

Comment:
So when any pynvml error happens, this method `_get_gpu_usage` will return an empty list `[]`. Ideally throwing exceptions would be a better design, but I don't think it's a good behavior to return an empty list. You may want to return `None` instead, to be exception-safe. Depends on how other `_get_xxx` methods are designed in https://github.com/ray-project/ray/blob/master/dashboard/modules/reporter/reporter_agent.py.

Reply:
Yes, that seems fair. IMHO null and empty collection should be distinguished well, but if that's the design and is consistent with other reporter fields, it should be OK. Sounds good to me!